download_data/scripts/generate_samplesheet.py

```diff
@@ -5,14 +5,73 @@
 import glob
 import os
 import sys
+import requests
+import xmltodict
+import json
 
+from ena_portal_api.ena_handler import EnaApiHandler
+handler = EnaApiHandler()
 
-def main(assembly_dir, run_dir, rename_file, output):
+def load_xml(assembly):
+    retry_attempts = 3
+    retry_delay_min = 15
+    xml_url = "https://www.ebi.ac.uk/ena/browser/api/xml/{}".format(assembly)
+
+    for attempt in range(1, retry_attempts + 1):
+        r = requests.get(xml_url)
+        if not r.ok:
+            retry_delay = retry_delay_min * attempt
+            if r.status_code == 500:
+                print(f"Received HTTP 500 error for sample {assembly}. Retrying in {retry_delay} seconds...")
+                time.sleep(retry_delay)
+                continue
+            else:
+                print(f"Unable to fetch xml for sample {assembly}. Retrying in {retry_delay} seconds...")
+                time.sleep(retry_delay)
+                continue
+        if r.ok:
+            data_dict = xmltodict.parse(r.content)
+            json_dump = json.dumps(data_dict)
+            json_data = json.loads(json_dump)
+            return json_data
+        else:
+            print("Could not retrieve xml for accession", assembly)
+            print(r.text)
+            return None
+
+
+def get_erz_list(input_dir):
+    erz_list = set()
+    for file in glob.glob(os.path.join(input_dir, 'ERZ*')):
+        acc = file.strip().split('/')[-1].split('.')[0]
+        erz_list.add(acc)
+    return erz_list
+
+
+def get_err_erz(erz_list, outfile_software):
     accessions = {}
-    with open(rename_file, 'r') as file_in:
-        for line in file_in:
-            line = line.strip().split('\t')  # err,erz
-            accessions[line[0]] = line[1]
+    with open(outfile_software, 'w') as file_software:
+        for erz_acc in erz_list:
+            handler_request = handler.get_assembly(erz_acc)
+            run_acc = handler_request["submitted_ftp"].strip().split('/')[-1].split('.')[0]
+            assembly_software = handler_request["assembly_software"]
+            print(assembly_software)
+            if not assembly_software:
+                json_analysis = load_xml(erz_acc)
+                assembly_software = json_analysis["ANALYSIS_SET"]["ANALYSIS"]["ANALYSIS_TYPE"]["SEQUENCE_ASSEMBLY"]["PROGRAM"]
+                assembly_software = assembly_software.replace(" ", "_v")
+                print(assembly_software)
+            if not run_acc.startswith(('ERR', 'DRR', 'SRR')):
```

Comment:
Don't think it's important to change this now, but you could use the `logging` library for printing error messages like this. It's functionally the same I'm pretty sure it just gives nicer formatting on messages:
```
import logging

logging.basicConfig(level=logging.DEBUG)

logging.error("'Invalid run name {} for assembly {}'.format(run_acc, erz_acc))
exit(1)
```